icu4c/source/common/bytesinkutil.h

```diff
@@ -43,6 +43,16 @@ class U_COMMON_API CharStringByteSink : public ByteSink {
     CharString& dest_;
 };
 
+// CharString doesn't provide the public API that StringByteSink requires a
+// string class to have so this template specialization replaces the default
+// implementation of StringByteSink<CharString> with CharStringByteSink.
+template<>
+class StringByteSink<CharString> : public CharStringByteSink {
```

Comment:
> I believe you, but it's surprising to me that a template specialization can switch the base class compared with the original definition...

If you think of the template specialization as a blank slate, instead of some kind of override, it might become more clear why this works (and that it's intentional that this should work).